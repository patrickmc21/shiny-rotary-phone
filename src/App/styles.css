.App {
  margin: 0;
  padding: 0;
  background-image: url('./images/background.jpg');
  background-size: cover;
}

.App-header {
  padding-top: 20px;
  height: 200px;
  box-sizing: border-box;
}

.App-title {
  margin: 0 auto;
  height: 150px;
  width: 320px;
  background-image: url('./images/logo.png');
  background-size: cover;
}

.return-home {
  float: right;
  padding: 3px;
  border: none;
  border: 1px solid #F3BF44;
  border-radius: 40px;
  border-style: outset;
  margin-right: 20px;
  width: 140px;
  background-color: rgba(0,0,0,0.4);
  box-shadow: 0px 0px 5px 1px #F3BF44;
  color: #F3BF44;
  font-family: 'Nanum Gothic', sans-serif;
  font-size: 28px;
  text-align: center;
  text-decoration: none;
}

.return-home:hover {
  border: 1px solid #9F1F16;
  box-shadow: 0px 0px 5px 1px #9F1F16;
  color: #9F1F16;
}

/*9F1F16 rebel red*/
/*36363D imperial grey*/