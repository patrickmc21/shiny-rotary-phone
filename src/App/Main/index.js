import React from 'react';
import ContentCard from './ContentCard/index.js';
import PropTypes from 'prop-types';
import './styles.css';

const Main = () => {
  return (
    <div>I'm the Main
      <ContentCard />
    </div>

  )
}

export default Main;