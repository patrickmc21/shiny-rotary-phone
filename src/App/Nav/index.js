import React from 'react';
import ContentButton from './ContentButton/index.js';
import Favorites from './Favorites/index.js';
import PropTypes from 'prop-types';
import './styles.css';

const Nav = () => {
  return (
    <div>I'm the Nav
      <ContentButton />
      <Favorites />
    </div>
  )
}

export default Nav;