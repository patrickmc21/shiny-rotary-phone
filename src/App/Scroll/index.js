import React from 'react';
import PropTypes from 'prop-types';
import './styles.css';

const Scroll = () => {
  return (
    <div>I'm the Scroll</div>
  )
}

export default Scroll;