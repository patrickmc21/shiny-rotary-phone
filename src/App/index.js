import React, { Component } from 'react';
import Nav from '../Nav/index.js';
import Scroll from '../Scroll/index.js';
import Main from '../Main/index.js';
import { Route, Link } from 'react-router-dom';
import PropTypes from 'prop-types';
import './styles.css';

class App extends Component {
  constructor(props) {
    super(props);
    this.state = {
      currentUser: '',
      scrollData: {},
      categories: {
        people: false, 
        vehicles: false,
        starships: false, 
        planets: false, 
        favorites: false
      },
      activeCategoryInfo: [],
      activeCategoryName: 'main',
      favorites: [],
      next: null,
      previous: null
    }
  }

  // <iframe width="560" height="315" src="https://www.youtube.com/embed/Aja1pCUZPso?rel=0&amp;controls=0&amp;showinfo=0" frameborder="0" allow="autoplay; encrypted-media" allowfullscreen></iframe>

  componentDidMount() {
    this.fetchScroll();
    if (localStorage.getItem('lastActiveUser')) {
      const lastActiveUser = localStorage.getItem('lastActiveUser');
      this.retrieveFavorites(lastActiveUser);
    }
  }

  fetchScroll = () => {
    const randomNumber = Math.floor(Math.random() * 7) + 1;
    const endpoint = `films/${randomNumber}`
    this.props.apiFetchCalls(endpoint)
      .then(scrollData=> this.setState({scrollData}))
      .catch(error => this.setState({errorStatus: error.message}))
  }

  userLogin = (name) => {
    const lastUser = localStorage.getItem('lastActiveUser');
    const userFavs = localStorage.getItem(name);
    if (lastUser === name || userFavs) {
      this.retrieveUser(name);
    } else {
      this.addUser(name);
    }
  }

  addUser = (name) => {
    localStorage.setItem('lastActiveUser', name);
    localStorage.setItem(name, JSON.stringify([]));
    let newFavorites = [];
    this.setState({
      favorites: newFavorites,
      currentUser: name
    });
  }

  retrieveUser = (name) => {
    localStorage.setItem('lastActiveUser', name);
    const userFavorites = JSON.parse(localStorage.getItem(name));
    this.setState({
      favorites: userFavorites,
      currentUser: name
    });
  }

  userLogOut= () => {
    this.setState({currentUser: ''});
  }

  retrieveFavorites = (name) => {
    const retrievedFavs = localStorage.getItem(name);
    const favorites = JSON.parse(retrievedFavs);
    this.setState({
      favorites,
      currentUser: name
    });
  }

  activateCategory = (category) => {
    const previousCategoryName = this.state.previousCategoryName;
    const currentCategories = {...this.state.categories};
    const newCategories = Object.keys(currentCategories).reduce((acc, type) => {
      type === category ? 
        acc[type] = true
        : acc[type] = false
      return acc;
    }, {})
    this.updateCurrentCategory(category)
    this.setState({
      categories: newCategories,
      activeCategoryName: category,
      previousCategoryName
    })
  }

  updateCurrentCategory = (category, endPoint) => {
    if (category !== 'favorites') {
      this.props.apiFetchCalls(`${category}/${endPoint || ''}`)
        .then(apiReturn => {
          const next = this.cleanEndPoint(category, apiReturn.next);
          const previous = this.cleanEndPoint(category, apiReturn.previous);
          this.setState({next, previous});
          return apiReturn;
      })
        .then(categoryInfo => this.props.swapiCleaners[category](categoryInfo.results))
        .then(activeCategoryInfo => this.setState({ activeCategoryInfo }))
        .catch(error => console.log(error.message))
    } else {
      this.setState({activeCategoryInfo: this.state.favorites})
    }
  }

  cleanEndPoint = (category, endPoint) => {
    if (endPoint) {
      const cleaned = endPoint.split(`${category}/`);
      return cleaned[1];
    }
  }

  addToFavorites = (favoriteObject) => {
    const currentUser = this.state.currentUser
    let favorites = [...this.state.favorites];
    if (!favorites.find(item => item.name === favoriteObject.name)) {
      favorites.push(favoriteObject);
    } else {
      favorites = favorites.filter(item => item.name !== favoriteObject.name);
    }
    localStorage.setItem(currentUser, JSON.stringify(favorites));
    this.setState({favorites})
  }

  render() {
    return (
      <div className="App">
        <Route exact path='/' render={() => {
          return (
            <Scroll 
              scrollData={this.state.scrollData}
              userLogin={this.userLogin}
              lastUser={this.state.currentUser}/>)
            }}/>
        <Route path={`/${this.state.activeCategoryName}`} render={() => {
          return (
            <div className={`wrapper ${this.state.activeCategoryName}`}>
              <header className="App-header">
                <Link 
                  to='/' 
                  className='return-home' 
                  onClick={this.userLogOut}>
                    Logout
                </Link>
                <div className="App-title"></div>
              </header>
              <Nav 
                activateCategory={this.activateCategory}
                buttonType={Object.keys(this.state.categories)}
                numberOfFavorites={this.state.favorites ? this.state.favorites.length : 0}
                activeCategoryName={this.state.activeCategoryName}/>
              <Main 
                activeCategoryInfo={this.state.activeCategoryInfo}
                activeCategoryName={this.state.activeCategoryName}
                updateCurrentCategory={this.updateCurrentCategory}
                addToFavorites={this.addToFavorites}
                favorites={this.state.favorites}
                next={this.state.next}
                previous={this.state.previous}/>
            </div>
          )
        }}/>

      </div>
    );
  }
}

App.propTypes = {
  apiFetchCalls: PropTypes.func,
  swapiCleaners: PropTypes.object
}

export default App;
