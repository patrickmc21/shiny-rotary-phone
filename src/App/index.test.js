import React from 'react';
import App from './index.js';
import { shallow, mount } from 'enzyme';

describe('App', () => {
  it('should match the snapshot', () => {
    const wrapper = shallow(<App />);
    expect(wrapper).toMatchSnapshot();
  })
})