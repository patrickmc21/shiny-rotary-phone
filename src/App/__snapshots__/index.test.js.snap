// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App should match the snapshot 1`] = `
ShallowWrapper {
  "length": 1,
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <App />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": Array [
        <header
          className="App-header"
>
          <h1
                    className="App-title"
          >
                    SWapiBox
          </h1>
</header>,
        <Nav />,
        <Scroll />,
        <Main />,
      ],
      "className": "App",
    },
    "ref": null,
    "rendered": Array [
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": <h1
            className="App-title"
>
            SWapiBox
</h1>,
          "className": "App-header",
        },
        "ref": null,
        "rendered": Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": "SWapiBox",
            "className": "App-title",
          },
          "ref": null,
          "rendered": "SWapiBox",
          "type": "h1",
        },
        "type": "header",
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "function",
        "props": Object {},
        "ref": null,
        "rendered": null,
        "type": [Function],
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "function",
        "props": Object {},
        "ref": null,
        "rendered": null,
        "type": [Function],
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "function",
        "props": Object {},
        "ref": null,
        "rendered": null,
        "type": [Function],
      },
    ],
    "type": "div",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": Array [
          <header
            className="App-header"
>
            <h1
                        className="App-title"
            >
                        SWapiBox
            </h1>
</header>,
          <Nav />,
          <Scroll />,
          <Main />,
        ],
        "className": "App",
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": <h1
              className="App-title"
>
              SWapiBox
</h1>,
            "className": "App-header",
          },
          "ref": null,
          "rendered": Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": "SWapiBox",
              "className": "App-title",
            },
            "ref": null,
            "rendered": "SWapiBox",
            "type": "h1",
          },
          "type": "header",
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "function",
          "props": Object {},
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "function",
          "props": Object {},
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "function",
          "props": Object {},
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
      ],
      "type": "div",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
      },
    },
  },
}
`;
